### PR TITLE
benchmark rpc ps

### PR DESCRIPTION
**Pull request by gcramer23**
DDP with NCCL AllReduce for the entire model experiment from Quip https://fb.quip.com/iQUtAeKIxWpF

I have been testing this on the AI cluster. There seem to be some connection problems with RPC when using multiple trainers or parameter servers.

```
Namespace(bconfig_id='3', dconfig_id='DummyData', mconfig_id='DummyModel', pconfig_id='None', tconfig_id='DdpNcclTrainer')

benchmark warmup done

metrics for trainer=0
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.972    | 0.097122   | 0.311644  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.58138 | 4.31439  | 0.00229848 | 0.0479424 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.4807  | 0.222566 | 0.0555432  | 0.235676  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.191488 | 3.54099 | 3.11694  | 0.557106   | 0.746395  |
+-----------------------------------+----------+---------+----------+------------+-----------+
metrics for trainer=1
+-----------------------------------+----------+---------+----------+-------------+------------+
| name                              |      min |     max |     mean |    variance |      stdev |
+===================================+==========+=========+==========+=============+============+
| backward_metric,backward          | 2.4617   | 2.59174 | 2.51196  | 0.000938276 | 0.0306313  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| batch_level_metric,batch_all      | 4.22605  | 4.71757 | 4.27921  | 0.00468424  | 0.0684415  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| foward_metric,forward_pass        | 0.807936 | 1.50118 | 0.846008 | 0.00601693  | 0.0775688  |
+-----------------------------------+----------+---------+----------+-------------+------------+
| hook_future_metric,nccl_allreduce | 0.108544 | 0.1536  | 0.11222  | 2.16726e-05 | 0.00465538 |
+-----------------------------------+----------+---------+----------+-------------+------------+
metrics for all trainer
+-----------------------------------+----------+---------+----------+------------+-----------+
| name                              |      min |     max |     mean |   variance |     stdev |
+===================================+==========+=========+==========+============+===========+
| backward_metric,backward          | 2.45248  | 4.18304 | 3.24198  | 0.584391   | 0.764455  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| batch_level_metric,batch_all      | 4.11955  | 4.71757 | 4.2968   | 0.00378467 | 0.0615197 |
+-----------------------------------+----------+---------+----------+------------+-----------+
| foward_metric,forward_pass        | 0.141312 | 1.50118 | 0.534287 | 0.128284   | 0.358167  |
+-----------------------------------+----------+---------+----------+------------+-----------+
| hook_future_metric,nccl_allreduce | 0.108544 | 3.54099 | 1.61458  | 2.5456     | 1.59549   |
+-----------------------------------+----------+---------+----------+------------+-----------+
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57454 benchmark rpc ps**

Differential Revision: [D28296175](https://our.internmc.facebook.com/intern/diff/D28296175)

### CODE REVIEW

File: benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:24]: This means if this is called before the CUDA ops are done, it will be perf impacts. Shall we make the synchronize API explicit? I mean 


```python
    def elapsed_time(self, rank):
        if not self.end.query():
            raise RuntimeError("......")
        return self.start.elapsed_time(self.end)

    def synchronize():
        ....
```

File: benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:25]: IIRC, this will fail if `self.start` and `self.end` are on different devices?

File: benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:12]: Do we need to pass in rank here, or can it be specified in the ctor? Besides, passing it here means that this is not overriding the abstract `record_start` in `MetricBase`?

File: benchmarks/distributed/rpc/ps/metrics/MetricBase.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/MetricBase.py:25]: API not used.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:52]: API doc usually don't include self, because users would use it as `instance. get_processed_metrics()`.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:46]: If this mean to be an API doc, it should be placed inside the function implementation. See the following example.

https://github.com/pytorch/pytorch/blob/1d3a9bff3c3256bd8aec4ed066119eb32883715c/torch/distributed/rpc/api.py#L454-L460

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:10]: If `overwrite_metrics=True` was never used in this PR. Shall we remove it from this one and add it back when needed in the future?

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:12]: this means we don't expect to use the same logger to log both CPU and CUDA metrics?

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/launcher.py:149]: IIUC, this is not used in this PR? If yes, let's remove and add it when we need it.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/launcher.py:40]: curious, why do we need an additional helper function instead of putting this to the metrics printer class implementation?

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:26]: If this the intention, I guess you can allow the user to directly pass in a metric class (not string) to the ctor. I.e.:

```python
def __init__(self, rank, metric_class):
    self.metric_class = metric_class

def add_metric(...):
    self.metrics[type][key] = self.metric_class(name)
```

And users can call the ctor with 

```python
logger = MetricsLogger(0, CUDAMetric)
```

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:16]: the API name already contains a metric, we can simplify the args to type, key, name

File: benchmarks/distributed/rpc/ps/trainers/TrainerBase.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/trainers/TrainerBase.py:24]: curious, is there any reason that need a different name. The call trace is like:

record_start (TrainerBase) -> add_metric (MetricLogger) -> record_start(Metric)

Will there be any issue if we all stay with the same name?

File: benchmarks/distributed/rpc/ps/trainers/DdpNcclTrainer.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/ps/trainers/DdpNcclTrainer.py:107]: I wonder if this training loop can be moved to the TrainerBase class?

Comment (mrshenli) [reply benchmarks/distributed/rpc/ps/trainers/DdpNcclTrainer.py:107]: Would I be correct if I assume different trainers only need different hook function and hook state, and all other code can be moved to TrainerBase?

File: benchmarks/distributed/rpc/ps/BenchmarkConfigurations.py
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/BenchmarkConfigurations.py:1]: python dataclass could simplify this class (https://docs.python.org/3/library/dataclasses.html)

File: benchmarks/distributed/rpc/ps/README.md
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/README.md:1]: nit: Change folder name from ps -> parameter_server? `.../rpc/parameter_server` sounds more intuitive than `.../rpc/ps` to me

File: benchmarks/distributed/rpc/ps/README.md
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/README.md:3]: This can be done at a later time. Perhaps creating a placeholder experiment and a quickstart of what to change in the placeholder experiment could better illustrate the steps needed

File: benchmarks/distributed/rpc/ps/bash_experiment_scripts/ddp_nccl_allreduce.sh
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/bash_experiment_scripts/ddp_nccl_allreduce.sh:13]: may want to note here or in the README that you need slurm to run this script

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:18]: can use f-strings here and below as well

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/launcher.py:24]: nit: use f-strings instead of .format. Same comment for rest of formatted strings in file

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/launcher.py:33]: Should we throw an error here instead?

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/launcher.py:325]: you can add the `help` parameter to provide information about each argument

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (H-Huang) [inline benchmarks/distributed/rpc/ps/launcher.py:247]: Don't we need a context manager when opening files? (e.g. `with open(file, "r") as f: ...` ), or does json.load handle that for you?

File: benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:24]: okay ill add the synchronize API.

File: benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:25]: Yes, it does fail. The API calls to `record_start` and `record_end` are from the logger. The logger is initialized with a device rank. So it shouldn't be a problem if a user is inheriting from a base class since they can't access the logger because of name mangling.

File: benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:12]: Well python does not call the parent method if a subclass has the method name. 

```
class A:

    def __init__(self):
        self.name = "A"

    def get_name(self):
        return self.name

class B(A):

    def __init__(self):
        super().__init__()
        self.name = "B"

    def get_name(self, name):
        return name

if __name__ == "__main__":
    b = B()
    # print(b.get_name("garrett"))
    print(b.get_name())
```

Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/CUDAMetric.py:12]: > can it be specified in the ctor

I think this is a better approach because it also deals with the issue of the metric recording on different devices. https://github.com/pytorch/pytorch/pull/57454#discussion_r625474043

File: benchmarks/distributed/rpc/ps/metrics/MetricBase.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricBase.py:25]: okay will remove it.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:52]: okay.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:46]: okay.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:10]: okay.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:12]: I was thinking about this. CPUMetrics are not used in the experiments that I am implementing, but I can update the logger to specify logging CUDA or CPU.

```python

    def cpu_record_start(self, type, key, name):
        ...

    def cuda_record_start(self, type, key, name):
        ...            

```

I'll update the existing wrapper in the base classes and add a CPU wrapper.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:149]: okay.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:40]: Ill move it to the metrics printer class.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:26]: I think just allowing the logger to log both is a better solution. https://github.com/pytorch/pytorch/pull/57454#discussion_r626144963

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:16]: okay.

File: benchmarks/distributed/rpc/ps/trainers/TrainerBase.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/trainers/TrainerBase.py:24]: no but adding the `cuda_record_start` and `cpu_record_start` the call trace will be

cuda_record_start (TrainerBase) -> cuda_record_start (MetricLogger) -> record_start(CUDAMetric)
cpu_record_start (TrainerBase) -> cpu_record_start (MetricLogger) -> record_start(CPUMetric)

File: benchmarks/distributed/rpc/ps/trainers/DdpNcclTrainer.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/trainers/DdpNcclTrainer.py:107]: It is hard to determine that all training loops will be identical. For example, maybe I want to call backward multiple times in my training loop.  

All trainers that inherit from DdpTrainerBase will have to implement their hook function and hook state. If a trainer is implemented that does not inherit from DdpTrainerBase, it does not need to implement a hook function and hook state. It will potentially have a different training loop that is not based on a communication hook.

I don't think the training loop should be in the base class.

File: benchmarks/distributed/rpc/ps/BenchmarkConfigurations.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/BenchmarkConfigurations.py:1]: okay I will use it.

File: benchmarks/distributed/rpc/ps/README.md
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/README.md:1]: okay.

File: benchmarks/distributed/rpc/ps/bash_experiment_scripts/ddp_nccl_allreduce.sh
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/bash_experiment_scripts/ddp_nccl_allreduce.sh:13]: ill add it to the script.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py:18]: okay.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:24]: okay.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:247]: No I use json to parse a file object.
```
f = open(file, "r")
json.load(f)
```

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:201]: okay.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:33]: Ill remove those two lines not needed.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:325]: okay.

File: benchmarks/distributed/rpc/ps/metrics/CPUMetric.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/CPUMetric.py:18]: Ill add a RuntimeError.

File: benchmarks/distributed/rpc/ps/models/DummyModel.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/models/DummyModel.py:8]: Ill remove the values.

File: benchmarks/distributed/rpc/ps/metrics/ProcessedMetricsPrinter.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/ProcessedMetricsPrinter.py:9]: okay.

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py:289]: There are some slight variations. I have added a helper function for loading the JSON.

File: benchmarks/distributed/rpc/ps/trainers/TrainerBase.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/trainers/TrainerBase.py:76]: Not in this pr, but I think this is a good decorator to have in the base class.

File: benchmarks/distributed/rpc/ps/README.md
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/README.md:3]: Okay, Ill make a note to do this later.

File: benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/metrics/MetricsLogger.py]: I ended up adding a CUDA parameter to indicate if the metric should be CUDA or CPU.
```python
    def record_start(self, type, key, name, cuda=True):
        self.__metrics_logger.record_start(
            type,
            key,
            name,
            cuda
        )
 ```

File: benchmarks/distributed/rpc/ps/trainers/TrainerBase.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/trainers/TrainerBase.py]: call trace is now record_start (TrainerBase) -> record_start (MetricLogger) -> record_start(Metric)

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (H-Huang) [reply benchmarks/distributed/rpc/ps/launcher.py]: Then we should update this to also close the file right? It is always best practice to close after opening to ensure resources (e.g. file descriptor) are released correctly. Your example above would be changed to:
```python
with open(file, "r") as f:
    json.load(f)
```

File: benchmarks/distributed/rpc/parameter_server/launcher.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/launcher.py:130]: nit: will it be better if `get_name` just take the `configurations` as an argument and then retrieve whatever fields it needs? It will help make the call site less verbose.

Comment (mrshenli) [reply benchmarks/distributed/rpc/parameter_server/launcher.py:130]: also in case there will be additional roles in the future, we don't need to update all call sites.

File: benchmarks/distributed/rpc/parameter_server/launcher.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/launcher.py:164]: curious, any reason that configurations, trainer_configurations, ps_configurations needs to be three different objects. The files can be separate, but can they all load into one `config` object and then functions using that config can just retrieve whatever they need?

File: benchmarks/distributed/rpc/parameter_server/launcher.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/launcher.py:256]: shall we change the hard-coded path and let users to pass in a file path using command line args?

Comment (mrshenli) [reply benchmarks/distributed/rpc/parameter_server/launcher.py:256]: this applies to other configurations as well.

File: benchmarks/distributed/rpc/parameter_server/launcher.py
Comment (mrshenli) [inline benchmarks/distributed/rpc/parameter_server/launcher.py:270]: after changing this to take a file path, let's also update the config names. We can call then --benchmark, --data, --model, --server, --trainer, for readability.

**Review by mrshenli — APPROVED** [review]
I left some minor comments. Stamp to unblock. Please consider addressing the minor comments before landing. I will create Github issues if I discover sth after landing. Thanks for working on this!

File: benchmarks/distributed/rpc/ps/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/ps/launcher.py]: yes forgot to close the file. Ill use close it.

File: benchmarks/distributed/rpc/parameter_server/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/launcher.py:130]: okay updated the get_name method.

File: benchmarks/distributed/rpc/parameter_server/launcher.py
Comment (gcramer23) [reply benchmarks/distributed/rpc/parameter_server/launcher.py:256]: okay updated.

**Comment by gcramer23** [comment]
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28296175).

**Review by H-Huang — APPROVED** [review]

**Pull request closed without merging**
